### PR TITLE
fix: Stop Markers and Popups from rendering on-screen when behind the camera

### LINKED ISSUE #8292: map.project() returns incorrect point for locations behind the camera + DOM Markers render in the sky
**maplibre-gl-js version**: 6.6.0 (reproduced on `main` at commit `1e9194fdb`)

**browser**: Chrome, Linux

AI helped me research this defect, build the reproduction, and drafted some of the text in this issue. 

### Steps to Trigger Behavior

1. Create a map with `pitch: 85` and `maxPitch: 85`, so that the horizon is visible on screen.
2. Add a `Marker`.
3. Move the map so the marker is behind the camera.

### Link to Demonstration

https://jsbin.com/yemazuf

This demo includes a DOM Marker, and a 'gl-point' rendered at the same location via a layer.  The gl-point is there to show that this issue only seems to exist for things in the DOM, and not things rendered via WebGL, but that is just my guessing, and not something that I dug deeply into.

 Press **Run the test** to see the bug demonstrated for you.

 The blue GL ring at the same location disappears, which is correct. The red DOM marker moves into the sky, which is the bug. The panel has some info in it that may be helpful, I'm not really sure. The agent dianosing this issue for me believes it's something to do with the "clip-space w".

My agent claims that this same issue is also present for `Popup`s as well, but I didn't include that in the demo.

### Expected Behavior

_Onscreen behavior_
A marker whose location is behind the camera should not be drawn on the map.

_`map.project()` behavior_
I'm not really sure what `map.project()` should actually return in this case.  Maybe just `Number.MAX_VALUE` for both the `x` and `y` of the marker? That should ensure it gets placed off screen.

### Actual Behavior

The marker element renders in the sky, above the horizon. 

### My Agents Root Cause

Like any good agent, mine confidently came up with a root cause .  This may be helpful for you 🤷 ...

<details>
<summary>Claude Opus root cause analysis</summary>

### root cause

`coordinatePoint()` divides by the clip-space `w` without checking its sign:

- `src/geo/projection/mercator_transform.ts:484-488` — builds a `vec4`, multiplies by the pixel matrix, then `return new Point(p[0] / p[3], p[1] / p[3]);`. Nothing tests `p[3]`.
- `src/geo/projection/mercator_transform.ts:362-366` — `locationToScreenPoint()` calls `coordinatePoint()`.
- `src/ui/map.ts:2027-2029` — `Map#project()` calls `locationToScreenPoint()`.
- `src/ui/marker.ts:752-793` — `_update` assigns `this._flatPos = this._pos = this._map.project(this._lngLat)._add(this._offset);` (line 762) and writes that straight into `this._element.style.transform` (line 789).

The transform layer already knows how to test the sign of `w` elsewhere: `projectTileCoordinates` at `src/geo/projection/mercator_transform.ts:878-894` returns `signedDistanceFromCamera: w` alongside the divided point, and symbol placement consumes it. So the information is available, it is just discarded on this path.

`Marker` has no compensating check. `_updateOpacity` (`src/ui/marker.ts:706`) asks `isLocationOccluded` (line 708), but `MercatorTransform.isLocationOccluded` returns `false` unconditionally (`src/geo/projection/mercator_transform.ts:854-856`), and there is no off-screen early return anywhere in that method.

### Scope

- `Popup` has the same gap. `Popup._updateOpacity` (`src/ui/popup.ts:281-290`) returns early when `options.locationOccludedOpacity` is `undefined`, which is the default, so a popup behind the camera gets no treatment at all.
- `VerticalPerspectiveTransform._projectSurfacePointToScreen` (`src/geo/projection/vertical_perspective_transform.ts:773-782`) divides by `projected[3]` with no sign check either. I do not believe it is reachable in the same way today: `GlobeTransform.locationToScreenPoint` (`src/geo/projection/globe_transform.ts:425-427`) delegates to `currentTransform`, which is the internal mercator transform when globe rendering is off. I have not tested the globe path, so I am flagging it rather than claiming it is broken.
- Separately, and I think this is its own defect: a marker covered by terrain is dimmed to `0.2` opacity (`src/ui/marker.ts:1082`) and gets the `maplibregl-marker-covered` class (line 713), but nothing in the stylesheet drops its pointer events, so it stays clickable while looking hidden. The class documentation at `src/ui/marker.ts:209-216` tells users to add `pointer-events: none` themselves. That path also depends on `Terrain.depthAtPoint` (`src/render/terrain.ts:429-437`), which does a synchronous `gl.readPixels` per marker.
- Tangentially, MapLibre issue #8184 "Map .transform helpers removal" is why user code now reaches through `map._camera.transform` for this kind of thing, as several lines above show.

</details>

### PR DESCRIPTION
**Pull request by schickm**
Fixes an issue where Markers and Popups continue to appear on screen after they have moved behind the camera. They end up being thrown up into the sky, but still on-screen.

The issue was solved by having `map.project()` return `Number.MAX_VALUE` for x/y when an item is behind the camera.

I considered `NaN` instead of `Number.MAX_VALUE`, but `NaN` that would break any math that folks would be doing with this value. So, this feels like the least intrusive option.

<table>
<tr><th>Before</th><th>After</th>
<tr><td>before-marker-in-the-sky</td><td>after-marker-hidden
</tr>
</table>

Closes #8292

## Launch Checklist

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Link to related issues.
 - [x] Include before/after visuals or gifs if this PR includes visual changes.
 - [x] Write tests for all new functionality.
 - [x] Document any changes to public APIs.
 - [x] If you changed code in a file that has a benchmark file next to it (`*.bench.ts`), post before/after results of `npm run bench` (the compare workflow is in `test/bench/README.md`).
 - [x] Add an entry to `CHANGELOG.md` under the `## main` section.
 - [x] Confirm you have read our AI policy [here](https://github.com/maplibre/maplibre/blob/main/AI_POLICY.md).

### CODE REVIEW

File: src/ui/marker.test.ts [hunk 1]
```diff
@@ -1079,6 +1079,24 @@ describe('marker', () => {
...
+        map.setZoom(10);
+        map.setCenter([0, 0]);
+        map.setPitch(80);
+
+        const marker = new Marker()
+            .setLngLat([0, -2])
+            .addTo(map);
+
+        expect(marker._pos.x).toBeGreaterThan(containerSize);
+        expect(marker._pos.y).toBeGreaterThan(containerSize);
```
Comment (CommanderStorm) [inline src/ui/marker.test.ts]: `containerSize` is only used here in the assertions but defined all the way at the top of the fn.
Intentional?

File: src/geo/projection/mercator_transform.test.ts [hunk 2]
```diff
@@ -446,6 +446,22 @@ describe('transform', () => {
...
+        transform.setPitch(80);
+
+        // In front of the camera, so the point projects normally.
+        expect(transform.locationToScreenPoint(new LngLat(0, 0.05)).y).toBeCloseTo(244.4, 1);
+
+        // Behind the camera plane, so no pixel corresponds to this location.
+        const behindCamera = transform.locationToScreenPoint(new LngLat(0, -2));
+        expect(behindCamera.x).toBe(Number.MAX_VALUE);
+        expect(behindCamera.y).toBe(Number.MAX_VALUE);
+    });
```
Comment (CommanderStorm) [inline src/geo/projection/mercator_transform.test.ts]: please split this into two tests, one in front of the camera and one behind.
This way you can move the comments to the description and they are thus unable to rot.

**Review by CommanderStorm — APPROVED** [review]
Makes overall sense to me.
One test is a bit weird and I'd ask you to have a look if this is entirely intentional the way you wrote

File: src/geo/projection/mercator_transform.test.ts (hunk 2 above)
Comment (schickm) [reply src/geo/projection/mercator_transform.test.ts]: Done!

File: src/ui/marker.test.ts (hunk 1 above)
Comment (schickm) [reply src/ui/marker.test.ts]: Good catch, I changed it so that it just reads the containerSize directly off the map.  That way the value doesn't drift if someone ever changes it for this test.

**Comment by schickm** [comment]
Here's the bench report for mercator_transform, ratio is current merge base compared to my branch:

|   benchmark    | ratio |
| ------------------|-------------|
| flat hit z2    | 1.06x |
| sloped hit z2  | 1.15x |
| sky miss z2    | 1.01x |
| flat hit z10   | 1.00x |
| sloped hit z10 | 1.10x |
| sky miss z10   | 1.01x |
| flat hit z16   | 1.08x |
| sloped hit z16 | 0.97x |
| sky miss z16   | 1.01x |

**Comment by schickm** [comment]
@CommanderStorm Everything looks good to go.  I'm assuming someone else will take over merging from here?  Thanks for the review!

File: src/ui/popup.test.ts [hunk 3]
```diff
@@ -375,6 +375,19 @@ describe('popup', () => {
...
     });
 
+    test('Popup whose location is behind the camera is not positioned inside the viewport', () => {
+        const map = createMap({maxPitch: 85, pitch: 80, zoom: 10, center: [0, 0]});
+        const popup = new Popup()
+            .setLngLat([0, -2])
+            .setText('Test')
+            .addTo(map);
+
+        expect(popup._pos.x).toBeGreaterThan(containerWidth);
```
Comment (HarelM) [inline src/ui/popup.test.ts]: This uses internal fields, better to use the public API

**Comment by HarelM** [comment]
Do you know if this was introduced recently? If so, maybe there's a different solution? 
Since max value might also cause math issues I'm not sure NaN is not a better approach, but I'm not sure. 
Can't we return the place off screen the marker should be at?
i.e. a place outside the screen?

**Comment by schickm** [comment]
@HarelM My agent sleuthing claims that this issue has been present since the fork from maplibre, so not a recently introduced issue.

You're right, max value has it's own math issues. `NaN` had some as well - specifically about folks using it to apply to CSS styling (which max value would have been fine for).  But, I can see the merits of just letting the position go in some natural direction of screen as one would expect.

Determining horizontally where to place it is straightforward.  But I wasn't able to land on a good way of knowing how far vertically off-screen to place it.  I went with just anchoring it to a full viewport height off screen.

Commits have been pushed!

**Comment by schickm** [comment]
@HarelM According to some agent-sleuthing, the code at fault (`coordinatePoint()` in `mercator_transform.ts`) has been around in the same shape since the mapbox fork. So it doesn't seem to be a recent introduction to me.

You're right about max value having math issues - it screws up things like `easeTo(...)`. I went with that approach because I thought it would be nice to be able to determine if something was behind the camera by just checking ` ... === Number.MAX_VALUE`.  But it does feel like an unexpected change for something that is just supposed to be projecting a lat/lng to a screen position.  I'm sure there are other ways to determine if something is offscreen cheaply.

> Can't we return the place off screen the marker should be at? i.e. a place outside the screen?

We can to a certain extent...Once a point is fully behind the camera we can determine horizontally where it would sit below the screen edge, however there isn't a way to determine how far below the screen edge to place it

File: src/geo/projection/mercator_transform.ts
Comment (HarelM) [inline src/geo/projection/mercator_transform.ts:1]: What about globe/vertical perspective transform?

File: src/geo/projection/mercator_transform.ts [hunk 4]
```diff
@@ -479,12 +479,35 @@ export class MercatorTransform implements ITransform {
...
+        return this._offScreenPointBehindCamera(p[0], p[1], w);
+    }
+
+    private _offScreenPointBehindCamera(x: number, y: number, w: number): Point {
+        const cx = this.width / 2;
+        const cy = this.height / 2;
+        const dx = x - cx * w;
+        let dy = y - cy * w;
+        if (dx === 0 && dy === 0) {
+            // Exactly behind the camera: no direction to leave through, so use straight down.
```
Comment (HarelM) [inline src/geo/projection/mercator_transform.ts]: Move comments to tsdoc of this method.

File: src/geo/projection/mercator_transform.test.ts
Comment (HarelM) [inline src/geo/projection/mercator_transform.test.ts:1]: Are all the added tests really needed? Can you consolidate some maybe and still keep the coverage of the newly added method?

**Comment by HarelM** [comment]
Added a few minor comments, overall looks good to me.

File: src/geo/projection/mercator_transform.ts
Comment (schickm) [reply src/geo/projection/mercator_transform.ts:1]: The same issue seems to be present in there as well. It seems that the fix might be a little different there.

I'm not using that transform for anything in my current work, and I'm already a little out of my depth, so I can't really justify spending company time on it right now. I'm happy to file an issue for it though so it's tracked as a gap in the codebase.

File: src/geo/projection/mercator_transform.ts (hunk 4 above)
Comment (schickm) [reply src/geo/projection/mercator_transform.ts]: fixed!

File: src/geo/projection/mercator_transform.test.ts
Comment (schickm) [reply src/geo/projection/mercator_transform.test.ts:1]: Good call, I consolidated them down and removed the super edge case one of "point is exactly behind the camera".  The setup for that one was too elaborate.

File: src/geo/projection/mercator_transform.ts
Comment (HarelM) [reply src/geo/projection/mercator_transform.ts:1]: Yes, please do that, hopefully with good showcase example. 
Although to be fair, it's probably one prompt to your favorite AI to check and update this support...

**Review by HarelM — APPROVED** [review]

File: src/geo/projection/mercator_transform.ts
Comment (HarelM) [reply src/geo/projection/mercator_transform.ts:1]: I'll let someone that find this issue annoying enough to report of fix. If there's no need for it right now, let's delay it...
